src/ReinforcementLearningCore/src/policies/agent/agent_base.jl

```diff
@@ -27,13 +27,31 @@ mutable struct Agent{P,T} <: AbstractPolicy
     end
 end
 
+mutable struct OfflineAgent{P,T,B,R} <: AbstractPolicy
+    policy::P
+    trajectory::T
+    behavior_agent::B #a behavior agent to fill the trajectory. Leave nothing if the trajectory is prefilled. Should share the same trajectory as the parent OfflineAgent
+    behavior_steps::Int #steps to fill the trajectory (defaults to capacity of )
+    behavior_reset_condition::R #the reset condition of the environment.
+    function OfflineAgent(policy::P, trajectory::T, behavior_agent = nothing, behavior_steps = ReinforcementLearningTrajectories.capacity(trajectory.container.traces), behavior_reset_condition = ResetAtTerminal()) where {P<:AbstractPolicy, T<:Trajectory}
+        if behavior_steps == Inf
+            @error "behavior_steps is infinite, please provide a finite integer."
+        end
+        agent = new{P,T, typeof(behavior_agent), typeof(behavior_reset_condition)}(policy, trajectory, behavior_agent, behavior_steps, behavior_reset_condition)
+        if TrajectoryStyle(trajectory) === AsyncTrajectoryStyle()
+            bind(trajectory, @spawn(optimise!(policy, trajectory)))
+        end
+        agent
+    end
+end
+
 Agent(;policy, trajectory) = Agent(policy, trajectory)
 
-RLBase.optimise!(agent::Agent, stage::S) where {S<:AbstractStage} = RLBase.optimise!(TrajectoryStyle(agent.trajectory), agent, stage)
-RLBase.optimise!(::SyncTrajectoryStyle, agent::Agent, stage::S) where {S<:AbstractStage} = RLBase.optimise!(agent.policy, stage, agent.trajectory)
+RLBase.optimise!(agent::Union{Agent,OfflineAgent}, stage::S) where {S<:AbstractStage} = RLBase.optimise!(TrajectoryStyle(agent.trajectory), agent, stage)
```

Comment:
This seems awkward…can we not have `OfflineAgent <: Agent` so we don’t need the union type? I suppose even having a new abstract type `AbstractAgent` if necessary would be preferable to the union type, right?